pytorch_lightning/loops/training_epoch_loop.py

```diff
@@ -86,12 +88,19 @@ def reset(self) -> None:
         # track epoch output
         self.epoch_output = [[] for _ in range(self.batch_loop.num_active_optimizers(self.total_batch_idx))]
 
+        # reset tracking
+        self.progress.reset_on_epoch()
+
     def on_run_start(self, *args: Any, **kwargs: Any) -> None:
+        self.progress.epoch.increment_ready()
+
         # hook
         self.trainer.logger_connector.on_epoch_start()
         self.trainer.call_hook("on_epoch_start")
         self.trainer.call_hook("on_train_epoch_start")
 
+        self.progress.epoch.increment_started()
```

Comment:
okay, so I see the pattern now I think. 

`increment_ready()` is always before we run hooks for the start phase.
`increment_started()` is always after the hook calling for the start phase.

... then runs the processing

`increment_processed` end of increment, but before we call hooks ending this phase.
`increment_completed` runs always after we call the hooks at the end of the phase.

is that the general pattern?

Reply:
https://docs.google.com/document/d/1U73b4uFML2MNdObh6MVsBpFDrgLUOU2Sd_x0_DVk-Bo/edit?usp=sharing

Reply:
Yes, exactly.